// react-native-ui-kitten/ui/input/input.component.js
/**
 * @license
 * Copyright Akveo. All Rights Reserved.
 * Licensed under the MIT License. See License.txt in the project root for license information.
 */
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import React from 'react';
import { StyleSheet, TextInput, TouchableWithoutFeedback, View, } from 'react-native';
import { Interaction, styled, } from '../../theme';
import { Text, } from '../text/text.component';
import { allWithRest, isValidString, } from '../support/services';
import { FlexStyleProps, } from '../support/typings';
/**
 * Styled Input component.
 *
 * @extends React.Component
 *
 * @method {() => void} focus - Requests focus for the given input or view. The exact behavior triggered
 * will depend on the platform and type of view.
 *
 * @method {() => void} blur - Removes focus from an input or view. This is the opposite of `focus()`.
 *
 * @method {() => boolean} isFocused - Returns if the input is currently focused.
 *
 * @method {() => void} clear - Removes all text from the input.
 *
 * @property {boolean} disabled - Determines whether component is disabled.
 * Default is `false`.
 *
 * @property {string} status - Determines the status of the component.
 * Can be `primary`, `success`, `info`, `warning` or `danger`.
 *
 * @property {string} size - Determines the size of the component.
 * Can be `giant`, `large`, `medium`, `small`, or `tiny`.
 * Default is `medium`.
 *
 * @property {string} label - Determines label of the component.
 *
 * @property {StyleProp<TextStyle>} labelStyle - Customizes label style.
 *
 * @property {string} caption - Determines caption of the component.
 *
 * @property {StyleProp<TextStyle>} captionStyle - Customizes caption style.
 *
 * @property {StyleProp<TextStyle>} textStyle - Customizes text style.
 *
 * @property {(style: StyleType) => React.ReactElement<ImageProps>} icon - Determines icon of the component.
 *
 * @property {(style: StyleType) => React.ReactElement<ImageProps>} captionIcon - Determines caption icon.
 *
 * @property TextInputProps
 *
 * @property StyledComponentProps
 *
 * @example Simple usage example
 *
 * ```
 * import React from 'react';
 * import { Input } from 'react-native-ui-kitten';
 *
 * export class InputShowcase extends React.Component {
 *
 *   public state = {
 *     inputValue: '',
 *   };
 *
 *   private onInputValueChange = (inputValue: string) => {
 *     this.setState({ inputValue });
 *   };
 *
 *   public render(): React.ReactNode {
 *     return (
 *       <Input
 *         value={this.state.inputValue}
 *         onChangeText={this.onInputValueChange}
 *       />
 *     );
 *   }
 * }
 * ```
 *
 * @example Inline styling example
 *
 * ```
 * import React from 'react';
 * import { Input, InputProps } from 'react-native-ui-kitten';
 *
 * export const InputShowcase = (props?: InputProps): React.ReactElement<InputProps> => {
 *   return (
 *     <Input
 *       style={styles.input}
 *       textStyle={styles.inputText}
 *       labelStyle={styles.inputLabel}
 *       captionStyle={styles.inputCaption}
 *       label='Label'
 *       caption='Caption'
 *       placeholder='Placeholder'
 *     />
 *   );
 * };
 * ```
 * */
export class InputComponent extends React.Component {
    constructor() {
        super(...arguments);
        this.textInputRef = React.createRef();
        this.focus = () => {
            this.textInputRef.current.focus();
        };
        this.blur = () => {
            this.textInputRef.current.blur();
        };
        this.isFocused = () => {
            return this.textInputRef.current.isFocused();
        };
        this.clear = () => {
            this.textInputRef.current.clear();
        };
        this.onFocus = (event) => {
            this.props.dispatch([Interaction.FOCUSED]);
            if (this.props.onFocus) {
                this.props.onFocus(event);
            }
        };
        this.onBlur = (event) => {
            this.props.dispatch([]);
            if (this.props.onBlur) {
                this.props.onBlur(event);
            }
        };
        this.onIconPress = (event) => {
            if (this.props.onIconPress) {
                this.props.onIconPress(event);
            }
        };
        this.getComponentStyle = (source) => {
            const flatStyles = StyleSheet.flatten(this.props.style);
            const _a = allWithRest(flatStyles, FlexStyleProps), { rest: inputContainerStyle } = _a, containerStyle = __rest(_a, ["rest"]);
            const { textMarginHorizontal, textFontSize, textLineHeight, textFontWeight, textColor, placeholderColor, iconWidth, iconHeight, iconMarginHorizontal, iconTintColor, labelColor, labelFontSize, labelLineHeight, labelMarginBottom, labelFontWeight, captionMarginTop, captionColor, captionFontSize, captionLineHeight, captionFontWeight, captionIconWidth, captionIconHeight, captionIconMarginRight, captionIconTintColor } = source, containerParameters = __rest(source, ["textMarginHorizontal", "textFontSize", "textLineHeight", "textFontWeight", "textColor", "placeholderColor", "iconWidth", "iconHeight", "iconMarginHorizontal", "iconTintColor", "labelColor", "labelFontSize", "labelLineHeight", "labelMarginBottom", "labelFontWeight", "captionMarginTop", "captionColor", "captionFontSize", "captionLineHeight", "captionFontWeight", "captionIconWidth", "captionIconHeight", "captionIconMarginRight", "captionIconTintColor"]);
            return {
                container: containerStyle,
                inputContainer: Object.assign({}, containerParameters, inputContainerStyle),
                captionContainer: {
                    marginTop: captionMarginTop,
                },
                text: {
                    marginHorizontal: textMarginHorizontal,
                    fontSize: textFontSize,
                    // FIXME: RN issue (https://github.com/facebook/react-native/issues/7823)
                    // lineHeight: textLineHeight,
                    fontWeight: textFontWeight,
                    color: textColor,
                },
                placeholder: {
                    color: placeholderColor,
                },
                icon: {
                    width: iconWidth,
                    height: iconHeight,
                    marginHorizontal: iconMarginHorizontal,
                    tintColor: iconTintColor,
                },
                label: {
                    color: labelColor,
                    fontSize: labelFontSize,
                    lineHeight: labelLineHeight,
                    marginBottom: labelMarginBottom,
                    fontWeight: labelFontWeight,
                },
                captionIcon: {
                    width: captionIconWidth,
                    height: captionIconHeight,
                    tintColor: captionIconTintColor,
                    marginRight: captionIconMarginRight,
                },
                captionLabel: {
                    fontSize: captionFontSize,
                    fontWeight: captionFontWeight,
                    lineHeight: captionLineHeight,
                    color: captionColor,
                },
            };
        };
        this.renderIconTouchableElement = (style) => {
            const iconElement = this.renderIconElement(style);
            return (<TouchableWithoutFeedback onPress={this.onIconPress}>
        {iconElement}
      </TouchableWithoutFeedback>);
        };
        this.renderIconElement = (style) => {
            const iconElement = this.props.icon(style);
            return React.cloneElement(iconElement, {
                key: 0,
                style: [style, styles.icon, iconElement.props.style],
            });
        };
        this.renderLabelElement = (style) => {
            return (<Text key={1} style={[style, styles.label, this.props.labelStyle]}>
        {this.props.label}
      </Text>);
        };
        this.renderCaptionElement = (style) => {
            return (<Text key={2} style={[style, styles.captionLabel, this.props.captionTextStyle]}>
        {this.props.caption}
      </Text>);
        };
        this.renderCaptionIconElement = (style) => {
            const iconElement = this.props.captionIcon(style);
            return React.cloneElement(iconElement, {
                key: 3,
                style: [style, styles.captionIcon, iconElement.props.style],
            });
        };
        this.renderComponentChildren = (style) => {
            const { icon, label, captionIcon, caption } = this.props;
            return [
                icon && this.renderIconTouchableElement(style.icon),
                isValidString(label) && this.renderLabelElement(style.label),
                isValidString(caption) && this.renderCaptionElement(style.captionLabel),
                captionIcon && this.renderCaptionIconElement(style.captionIcon),
            ];
        };
    }
    render() {
        const _a = this.props, { themedStyle, textStyle, disabled } = _a, restProps = __rest(_a, ["themedStyle", "textStyle", "disabled"]);
        const componentStyle = this.getComponentStyle(themedStyle);
        const [iconElement, labelElement, captionElement, captionIconElement,] = this.renderComponentChildren(componentStyle);
        return (<View style={[componentStyle.container, styles.container]}>
        {labelElement}
        <View style={[componentStyle.inputContainer, styles.inputContainer]}>
          <TextInput ref={this.textInputRef} {...restProps} style={[componentStyle.text, styles.text, textStyle]} placeholderTextColor={componentStyle.placeholder.color} editable={!disabled} onFocus={this.onFocus} onBlur={this.onBlur}/>
          {iconElement}
        </View>
        <View style={[componentStyle.captionContainer, styles.captionContainer]}>
          {captionIconElement}
          {captionElement}
        </View>
      </View>);
    }
}
InputComponent.styledComponentName = 'Input';
const styles = StyleSheet.create({
    container: {},
    inputContainer: {
        flexDirection: 'row',
        alignItems: 'center',
        width: '100%',
    },
    captionContainer: {
        flexDirection: 'row',
        alignItems: 'center',
    },
    text: {
        flexGrow: 1,
        flexShrink: 1,
        flexBasis: 'auto',
    },
    placeholder: {},
    icon: {},
    label: {},
    captionIcon: {},
    captionLabel: {},
});
export const Input = styled(InputComponent);
