// react-native-ui-kitten/ui/support/components/arrow/arrow.component.js
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import React from 'react';
import { StyleSheet, View, } from 'react-native';
export class Arrow extends React.Component {
    constructor() {
        super(...arguments);
        this.getComponentStyle = (source) => {
            const flatStyle = StyleSheet.flatten(source);
            return {
                container: {
                    // @ts-ignore: `width` is restricted to be a number
                    borderLeftWidth: flatStyle.width,
                    // @ts-ignore: `width` is restricted to be a number
                    borderRightWidth: flatStyle.width,
                    // @ts-ignore: `height` is restricted to be a number
                    borderBottomWidth: flatStyle.height,
                    borderBottomColor: flatStyle.backgroundColor,
                    borderLeftColor: 'transparent',
                    borderRightColor: 'transparent',
                    backgroundColor: 'transparent',
                },
            };
        };
    }
    render() {
        const _a = this.props, { style } = _a, props = __rest(_a, ["style"]);
        const componentStyle = this.getComponentStyle(style);
        return (<View {...props} style={[style, styles.container, componentStyle.container]}/>);
    }
}
const styles = StyleSheet.create({
    container: {},
});
