// react-native-ui-kitten/ui/support/components/tabIndicator/tabIndicator.component.js
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import React from 'react';
import { Animated, Easing, } from 'react-native';
export class TabIndicator extends React.Component {
    constructor() {
        super(...arguments);
        this.contentOffset = new Animated.Value(0);
        this.onContentOffsetAnimationStateChanged = (state) => {
            // no-op
        };
        this.onContentOffsetAnimationStateEnd = (result) => {
            // no-op
        };
        this.createOffsetAnimation = (params) => {
            const animationDuration = params.animated ? this.props.animationDuration : 0;
            return Animated.timing(this.contentOffset, {
                toValue: params.offset,
                duration: animationDuration,
                easing: Easing.linear,
            });
        };
        this.onLayout = (event) => {
            this.indicatorWidth = event.nativeEvent.layout.width;
            this.scrollToOffset({
                offset: this.indicatorWidth * this.props.selectedPosition,
                animated: false,
            });
        };
        this.getComponentStyle = () => {
            const widthPercent = 100 / this.props.positions;
            return {
                width: `${widthPercent}%`,
                // @ts-ignore: RN has no types for `Animated` styles
                transform: [{ translateX: this.contentOffset }],
            };
        };
    }
    componentDidMount() {
        this.contentOffset.addListener(this.onContentOffsetAnimationStateChanged);
    }
    shouldComponentUpdate(nextProps) {
        return this.props.selectedPosition !== nextProps.selectedPosition;
    }
    componentDidUpdate() {
        const { selectedPosition: index } = this.props;
        this.scrollToIndex({
            index,
            animated: true,
        });
    }
    componentWillUnmount() {
        this.contentOffset.removeAllListeners();
    }
    /**
     * scrolls indicator to passed index
     *
     * @param params (object) - {
     *  index: number,
     *  animated: boolean | undefined
     * }
     */
    scrollToIndex(params) {
        const { index } = params, rest = __rest(params, ["index"]);
        const offset = this.indicatorWidth * index;
        this.scrollToOffset(Object.assign({ offset }, rest));
    }
    /**
     * scrolls indicator to passed offset
     *
     * @param params (object) - {
     *  offset: number,
     *  animated: boolean | undefined
     * }
     */
    scrollToOffset(params) {
        this.createOffsetAnimation(params).start(this.onContentOffsetAnimationStateEnd);
    }
    render() {
        const _a = this.props, { style } = _a, restProps = __rest(_a, ["style"]);
        const componentStyle = this.getComponentStyle();
        return (<Animated.View {...restProps} onLayout={this.onLayout} style={[style, componentStyle]}/>);
    }
}
TabIndicator.defaultProps = {
    selectedPosition: 0,
    animationDuration: 200,
};
