// react-native-ui-kitten/ui/checkbox/checkbox.component.js
/**
 * @license
 * Copyright Akveo. All Rights Reserved.
 * Licensed under the MIT License. See License.txt in the project root for license information.
 */
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import React from 'react';
import { StyleSheet, TouchableOpacity, View, } from 'react-native';
import { Interaction, styled, } from '../../theme';
import { Text, } from '../text/text.component';
import { CheckMark } from '../support/components';
import { isValidString } from '../support/services';
/**
 * Styled CheckBox component.
 *
 * @extends React.Component
 *
 * @property {boolean} checked - Determines whether component is checked.`
 * Default is `false`.
 *
 * @property {boolean} disabled - Determines whether component is disabled.
 * Default is `false.
 *
 * @property {string} status - Determines the status of the component.
 * Can be `primary`, `success`, `info`, `warning` or `danger`.
 *
 * @property {string} text - Determines text of the component.
 *
 * @property {StyleProp<TextStyle>} textStyle - Customizes text style.
 *
 * @property {(checked: boolean) => void} onChange - Fires on checkbox value change.
 *
 * @property TouchableOpacityProps
 *
 * @property StyledComponentProps
 *
 * @example Simple usage example
 *
 * ```
 * import React from 'react';
 * import { CheckBox } from 'react-native-ui-kitten';
 *
 * export class CheckBoxShowcase extends React.Component {
 *
 *   public state = {
 *     checked: false,
 *   };
 *
 *   private onChange = (checked: boolean) => {
 *     this.setState({ checked });
 *   };
 *
 *   public render(): React.ReactNode {
 *     return (
 *       <CheckBox
 *         checked={this.state.checked}
 *         onChange={this.onChange}
 *       />
 *     );
 *   }
 * }
 * ```
 *
 * @example Inline styling example
 *
 * ```
 * import React from 'react';
 * import { CheckBox, CheckBoxProps } from 'react-native-ui-kitten';
 *
 * export const CheckBoxShowcase = (props?: CheckBoxProps): React.ReactElement<CheckBoxProps> => {
 *   return (
 *     <CheckBox
 *       style={styles.checkbox}
 *       textStyle={styles.checkboxText}
 *       text='Place your text'
 *       checked={this.state.checked}
 *     />
 *   );
 * };
 * ```
 * */
class CheckBoxComponent extends React.Component {
    constructor() {
        super(...arguments);
        this.onPress = () => {
            this.props.dispatch([]);
            if (this.props.onChange) {
                this.props.onChange(!this.props.checked, false);
            }
        };
        this.onPressIn = (event) => {
            this.props.dispatch([Interaction.ACTIVE]);
            if (this.props.onPressIn) {
                this.props.onPressIn(event);
            }
        };
        this.onPressOut = (event) => {
            this.props.dispatch([]);
            if (this.props.onPressOut) {
                this.props.onPressOut(event);
            }
        };
        this.getComponentStyle = (source) => {
            const { textMarginHorizontal, textColor, textFontSize, textFontWeight, textLineHeight, iconWidth, iconHeight, iconBorderRadius, iconTintColor, outlineWidth, outlineHeight, outlineBorderRadius, outlineBackgroundColor } = source, containerParameters = __rest(source, ["textMarginHorizontal", "textColor", "textFontSize", "textFontWeight", "textLineHeight", "iconWidth", "iconHeight", "iconBorderRadius", "iconTintColor", "outlineWidth", "outlineHeight", "outlineBorderRadius", "outlineBackgroundColor"]);
            return {
                container: {},
                highlightContainer: {},
                selectContainer: containerParameters,
                text: {
                    marginHorizontal: textMarginHorizontal,
                    color: textColor,
                    fontSize: textFontSize,
                    fontWeight: textFontWeight,
                    lineHeight: textLineHeight,
                },
                icon: {
                    width: iconWidth,
                    height: iconHeight,
                    borderRadius: iconBorderRadius,
                    backgroundColor: iconTintColor,
                },
                highlight: {
                    width: outlineWidth,
                    height: outlineHeight,
                    borderRadius: outlineBorderRadius,
                    backgroundColor: outlineBackgroundColor,
                },
            };
        };
        this.createHitSlopInsets = (style) => {
            const { width } = StyleSheet.flatten(style);
            // @ts-ignore: `width` is restricted to be a number
            const value = 40 - width;
            return {
                left: value,
                top: value,
                right: value,
                bottom: value,
            };
        };
        this.renderTextElement = (style) => {
            const { text, textStyle } = this.props;
            return (<Text style={[style, styles.text, textStyle]}>{text}</Text>);
        };
        this.renderSelectIconElement = (style) => {
            return (<CheckMark style={[style, styles.icon]}/>);
        };
        this.renderIndeterminateIconElement = (style) => {
            return (<View style={[style, styles.icon]}/>);
        };
        this.renderIconElement = (style) => {
            if (this.props.indeterminate) {
                return this.renderIndeterminateIconElement(style);
            }
            else {
                return this.renderSelectIconElement(style);
            }
        };
        this.renderComponentChildren = (style) => {
            const { text } = this.props;
            return [
                this.renderIconElement(style.icon),
                isValidString(text) && this.renderTextElement(style.text),
            ];
        };
    }
    render() {
        const _a = this.props, { themedStyle, style, disabled, text } = _a, derivedProps = __rest(_a, ["themedStyle", "style", "disabled", "text"]);
        const _b = this.getComponentStyle(themedStyle), { container, highlightContainer, highlight, selectContainer } = _b, componentStyle = __rest(_b, ["container", "highlightContainer", "highlight", "selectContainer"]);
        const selectContainerStyle = [selectContainer, styles.selectContainer];
        const hitSlopInsets = this.createHitSlopInsets(selectContainerStyle);
        const [iconElement, textElement] = this.renderComponentChildren(componentStyle);
        return (<TouchableOpacity style={[container, styles.container, style]} activeOpacity={1.0} disabled={disabled} hitSlop={hitSlopInsets} onPress={this.onPress} onPressIn={this.onPressIn} onPressOut={this.onPressOut}>
        <View style={[highlightContainer, styles.highlightContainer]}>
          <View style={[highlight, styles.highlight]}/>
          <TouchableOpacity activeOpacity={1.0} {...derivedProps} disabled={disabled} style={selectContainerStyle} onPress={this.onPress} onPressIn={this.onPressIn} onPressOut={this.onPressOut}>
            {iconElement}
          </TouchableOpacity>
        </View>
        {textElement}
      </TouchableOpacity>);
    }
}
CheckBoxComponent.styledComponentName = 'CheckBox';
const styles = StyleSheet.create({
    container: {
        flexDirection: 'row',
        alignItems: 'center',
    },
    highlightContainer: {
        justifyContent: 'center',
        alignItems: 'center',
    },
    selectContainer: {
        justifyContent: 'center',
        alignItems: 'center',
    },
    highlight: {
        position: 'absolute',
    },
    icon: {},
    text: {},
});
export const CheckBox = styled(CheckBoxComponent);
